mullvad-relay-selector/src/relay_selector/query.rs

```diff
@@ -0,0 +1,779 @@
+///
+/// Creating a basic `RelayQuery` to filter relays by location, ownership and tunnel protocol:
+///
+/// ```rust
+/// // Create a query for a Wireguard relay that is owned by Mullvad and located in Norway.
+/// // The endpoint should specify port 443.
+/// use mullvad_relay_selector::query::RelayQuery;
+/// use mullvad_relay_selector::query::builder::RelayQueryBuilder;
+/// use mullvad_relay_selector::query::builder::{Ownership, GeographicLocationConstraint};
+///
+/// let query: RelayQuery = RelayQueryBuilder::new()
+///     .wireguard()                                            // Specify the tunnel protocol
+///     .location(GeographicLocationConstraint::country("no"))  // Specify the country as Norway
+///     .ownership(Ownership::MullvadOwned)                     // Specify that the relay must be owned by Mullvad
+///     .port(443)                                              // Specify the port to use when connecting to the relay
+///     .build();                                               // Construct the query
+/// ```
+///
+/// This example demonstrates creating a `RelayQuery` which can then be passed
+/// to the [`crate::RelaySelector`] to find a relay that matches the criteria.
+/// See [`builder`] for more info on how to construct queries.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RelayQuery {
+    pub location: Constraint<LocationConstraint>,
+    pub providers: Constraint<Providers>,
+    pub ownership: Constraint<Ownership>,
+    pub tunnel_protocol: Constraint<TunnelType>,
+    pub wireguard_constraints: WireguardRelayQuery,
+    pub openvpn_constraints: OpenVpnRelayQuery,
+}
+
+impl RelayQuery {
+    /// Create a new [`RelayQuery`] with no opinionated defaults. This
+    /// should be the const equivalent to [`Default::default`].
+    pub const fn new() -> RelayQuery {
+        RelayQuery {
+            location: Constraint::Any,
+            providers: Constraint::Any,
+            ownership: Constraint::Any,
+            tunnel_protocol: Constraint::Any,
+            wireguard_constraints: WireguardRelayQuery::new(),
+            openvpn_constraints: OpenVpnRelayQuery::new(),
+        }
+    }
+}
+
+impl Intersection for RelayQuery {
+    /// `intersection` defines a cautious merge strategy between two
+    /// [`RelayQuery`].
+    ///
+    /// * If two [`RelayQuery`] differ in any configuration such that no
+    /// consensus can be reached, the two [`RelayQuery`] are said to be
+    /// incompatible and `intersection` returns [`Option::None`].
+    ///
+    /// * Otherwise, a new [`RelayQuery`] is returned where each constraint is
+    /// as specific as possible. See [`Constraint`] for further details.
+    ///
+    /// This way, if the mullvad app wants to check if the user's configured
+    /// [`RelayQuery`] are compatible with any other [`RelayQuery`], taking the
+    /// intersection between them will never result in a situation where the app
+    /// can override the user's preferences.
```

Comment:
```suggestion
    /// Return a new query for which matches each relay that also matches both the intersected
    /// queries, and no other relay.
    ///
    /// * If two [`RelayQuery`]s differ such that no relay matches both, [`Option::None`] is
    ///   returned:
    /// ```rust
    /// # use mullvad_relay_selector::query::builder::RelayQueryBuilder;
    /// # use crate::mullvad_relay_selector::query::Intersection;
    /// let query_a = RelayQueryBuilder::new().wireguard().build();
    /// let query_b = RelayQueryBuilder::new().openvpn().build();
    /// assert_eq!(query_a.intersection(query_b), None);  
    /// ```
    ///
    /// * Otherwise, a new [`RelayQuery`] is returned where each constraint is
    /// as specific as possible. See [`Constraint`] for further details.
    ///
    /// This way, if the mullvad app wants to check if the user's configured
    /// [`RelayQuery`] are compatible with any other [`RelayQuery`], taking the
    /// intersection between them will never result in a situation where the app
    /// can override the user's preferences.
```

Reply:
It would also be super neat if you could add this to the second bullet point:

```rust
# use crate::mullvad_relay_selector::query::Intersection;
let query_a = RelayQueryBuilder::new().wireguard().build();
let query_b = RelayQueryBuilder::new().GeographicLocationConstraint::city("got").build();
let relay = todo!("se-got-wg-001");
assert_eq!(query_a.intersection(query_b).matches(relay), true);  
```

Reply:
Very nice! Thanks :blush: